RecoLocalCalo/HcalRecProducers/src/HFPreReconstructor.cc

```diff
@@ -0,0 +1,300 @@
+        const HFQIE10Info* info = &qie10Infos_[0];
+        for (unsigned i=0; i<sz; ++i)
+        {
+            const HcalDetId id(info[i].id());
+            sortedQIE10Infos_.push_back(QIE10InfoWithId(PmtAnodeId(id.baseDetId(), id.depth()), info+i));
+        }
+        std::sort(sortedQIE10Infos_.begin(), sortedQIE10Infos_.end());
+
+        // Count the PMTs
+        HcalDetId previousBaseId(sortedQIE10Infos_[0].first.first);
+        pmtCount = 1;
+        for (unsigned i=1; i<sz; ++i)
+        {
+            const HcalDetId baseId(sortedQIE10Infos_[i].first.first);
+            if (baseId != previousBaseId)
+            {
+                previousBaseId = baseId;
+                ++pmtCount;
+            }
+        }
+    }
+    return pmtCount;
+}
+
+void
+HFPreReconstructor::fillInfos(const edm::Event& e, const edm::EventSetup& eventSetup)
+{
+    using namespace edm;
+
+    // Clear the collection we want to fill in this method
+    qie10Infos_.clear();
+
+    // Get the Hcal topology if needed
+    ESHandle<HcalTopology> htopo;
+    if (tsFromDB_)
+    {
+        eventSetup.get<HcalRecNumberingRecord>().get(htopo);
+        paramTS_->setTopo(htopo.product());
+    }
+
+    // Get the calibrations
+    ESHandle<HcalDbService> conditions;
+    eventSetup.get<HcalDbRecord>().get(conditions);
+
+    // Get the input collection
+    Handle<QIE10DigiCollection> digi;
+    e.getByToken(tok_hfQIE10_, digi);
+
+    const unsigned inputSize = digi->size();
+    if (inputSize)
+    {
+        // Process the digis and fill out the HFQIE10Info vector
+        qie10Infos_.reserve(inputSize);
+
+        for (QIE10DigiCollection::const_iterator it = digi->begin();
+             it != digi->end(); ++it)
+        {
+            const QIE10DataFrame& frame(*it);
+
+            // Check zero suppression
+            if (dropZSmarkedPassed_)
+                if (frame.zsMarkAndPass())
+                    continue;
+
+            const HcalDetId cell(it->id());
+            const HcalCalibrations& calibrations(conditions->getHcalCalibrations(cell));
+            const HcalQIECoder* channelCoder = conditions->getHcalCoder(cell);
+            const HcalQIEShape* shape = conditions->getHcalShape(channelCoder);
+            const HcalCoderDb coder(*channelCoder, *shape);
+
+            int tsToUse = firstSample_;
+            if (tsFromDB_)
+            {
+                const HcalRecoParam* param_ts = paramTS_->getValues(cell.rawId());
+                tsToUse = param_ts->firstSample();
+            }
+
+            // Reconstruct the charge, energy, etc
+            qie10Infos_.push_back(reco_.reconstruct(frame, tsToUse, coder, calibrations));
+        }
+    }
+}
+
+void
+HFPreReconstructor::beginRun(const edm::Run& r, const edm::EventSetup& es)
+{
+    if (tsFromDB_)
+    {
+        edm::ESHandle<HcalRecoParams> p;
+        es.get<HcalRecoParamsRcd>().get(p);
+        paramTS_ = std::make_unique<HcalRecoParams>(*p.product());
```

Comment:
As far as I know, this is not expected to change during a run.
All other HCAL reco modules fetch this information inside
the "beginRun" method.

On 05/05/2016 03:56 AM, Slava Krutelyov wrote:
>
> In RecoLocalCalo/HcalRecProducers/src/HFPreReconstructor.cc <https://github.com/cms-sw/cmssw/pull/14383#discussion_r62163009>:
>
> > +            }
> > +
> > +            // Reconstruct the charge, energy, etc
> > +            qie10Infos_.push_back(reco_.reconstruct(frame, tsToUse, coder, calibrations));
> > +        }
> > +    }
> > +}
> > +
> > +void
> > +HFPreReconstructor::beginRun(const edm::Run& r, const edm::EventSetup& es)
> > +{
> > +    if (tsFromDB_)
> > +    {
> > +        edm::ESHandle<HcalRecoParams> p;
> > +        es.get<HcalRecoParamsRcd>().get(p);
> > +        paramTS_ = std::make_unique<HcalRecoParams>(*p.product());
>
> this will never ever change during a run, right (not even a relocation of the product as a result of upstream IOV changes) ?
> It may be better to move this to beginLumi.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub <https://github.com/cms-sw/cmssw/pull/14383/files/bb7a664cc1f3adac3512d57284b42aa643511ce5#r62163009>
>